default/python/AI/universe/system_network.py

```diff
@@ -92,3 +92,14 @@ def systems_connected_to_system(system_id: SystemId) -> Set[SystemId]:
         to_visit.extend(get_neighbors(current_system))
 
     return connected_systems
+
+
+@cache_for_current_turn
+def within_n_jumps(system_id: SystemId, n: int) -> FrozenSet[SystemId]:
+    if n < 1:
+        return frozenset({system_id})
+    inner_tier = within_n_jumps(system_id, n - 1)
+    result = set(inner_tier)
+    for sys_id in inner_tier:
+        result.update(get_neighbors(sys_id))
```

Comment:
We will call `get_neighbors` for system_id n times.  For each neighbor n-1 times.   
We use memorization for `get_neighbors`, so it won't affect performance very much.

Some experiments:
If we see 11 systems around it will be 90 calls
When we see 78 it's will 354 calls.

Classical BFS without recursion with check of visited systems will be a bit faster. 

It still pretty fast and 10 should be extream number.

Reply:
Well, at least duplicates can be avoided, it should be a little better now.

Reply:
I mean something like that,

```python
def within_n_jumps(system_id: SystemId, n: int) -> FrozenSet[SystemId]:
    current_layer = [system_id]
    next_layer = []
    visited = {system_id}

    for _ in range(n):
        for sid in current_layer:
            neighbors = get_neighbors(sid)
            not_visited = neighbors - visited
            next_layer.extend(not_visited)
            visited.update(not_visited)
        current_layer = next_layer
        next_layer = []

    return frozenset(visited)
```

When I was preparing for a job interview, I spent a lot of time solving tasks like that.

Reply:
Your version is indeed a little fast then mine, maybe 20%, but I had to create an extra big universe to measure that.

In fact when intermediate results are needed, my version is faster overall, thanks to caching. And I do need the results for different numbers when comparing supply effects of different species.

Reply:
Lets get back to this a bit later, I\ll merge it as is now.